src/order/extension/well.lean

```diff
@@ -0,0 +1,36 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Junyan Xu
+-/
+import data.prod.lex
+import set_theory.ordinal.arithmetic
+
+/-!
+# Extend a well-founded order to a well-order
+
+This file constructs a well-order (linear well-founded order) which is an extension of a given
+well-founded order.
+-/
+
+universe u
+
+variables {α : Type u} {r : α → α → Prop}
+
+namespace well_founded
+variable (hwf : well_founded r)
+include hwf
+
+/-- An arbitrary well order on `α` that extends `r`. -/
+noncomputable def well_order_extension : linear_order α :=
+let l : linear_order α := is_well_order.linear_order well_ordering_rel in by exactI
+  @linear_order.lift' α (ordinal ×ₗ α) _ (λ a : α, (hwf.rank a, a)) (λ _ _, congr_arg prod.snd)
```

Comment:
I'd like a little bit more documentation on how this works specifically. For example, that `l` is not used explicitly in the rest of the definition but that it's crucial that it's a well order.

Reply:
Maybe it also deserves its own definition, that you can extend one order by using another as a tie-breaker.

Reply:
Do you mean introduce a type synonym as for linear extensions?

Reply:
> I'd like a little bit more documentation on how this works specifically. For example, that l is not used explicitly in the rest of the definition but that it's crucial that it's a well order.

The proof was taken from https://math.stackexchange.com/a/4124696/12932

Essentially you maps an order into two linear (well) orders, one non-injective but respects the order, the other injective but doesn't respect the order. The lex product of the two well orders is still a well order, and the map to the product is injective and respects the order, so when you pull it back you get a well order extending the original order. It's a rather special construction and I'm not sure how much it can be generalized to make a separate definition.

Reply:
Oh the failure to infer `has_lt` turns out to be a universe issue, which is fixed by:
```suggestion
  @linear_order.lift' α (ordinal ×ₗ α) _
    (λ a : α, (well_founded.rank.{u u} hwf a, a)) (λ _ _, congr_arg prod.snd)
```
You may want to restrict the universe of `ordinal` in `acc.rank`, but there may be advantage to keep it polymorphic.

Reply:
Oh now I remember I already solved this problem once before deciding against the type synonym 🤦